[Macros] Use source locations to determine whether to suppress macro expansions.

hborla opened the pull request:
This PR is a follow-up to https://github.com/apple/swift/pull/64968 that replaces the request evaluator stack approach to macro expansion suppression in macro arguments with a source-location-based approach using ASTScope.

### Background

SE-0389 states that code in a macro argument does not have [visibility into macro-generated names](https://github.com/apple/swift-evolution/blob/main/proposals/0389-attached-macros.md#visibility-of-names-used-and-introduced-by-macros) for macros at the same scope or enclosing scopes, because the semantics of the program would depend on the ordering of those macro expansions. Arbitrary macro names also make name lookup extremely prone to request cycles, because any lookup into that scope will need to expand macros that can produce arbitrary names. For these reasons, macro resolution cannot be dependent on macro expansion, so macro-generated names are excluded from name lookup when lookup happens within a macro argument.

### The old request-evaluator-stack approach

Previously, answering the question of "are we performing name lookup from within a macro argument?" used a heuristic that checked whether there is an active `ResolveMacroRequest` in the current request-evaluator stack. This heuristic was not totally correct. For example, if type checking a macro argument was the first bit of type checker work that invoked conformance checking for a type, name lookups performed by the conformance checker would exclude macro-generated names even though the conforming type is not inside of a macro argument. If the conformance checker happens to be invoked before type checking the macro argument, macro-generated names would be included. Because macro-generated names in a conforming type can fulfill protocol requirements, this makes the type checking semantics dependent on the order of requests, which is Not Good.

### The new `ASTScope` approach

Instead, this change uses source locations to answer the question of "are we performing name lookup from within a macro argument?". Source locations are now plumbed through all of the name lookup APIs. When constructing the various name lookup requests, the source location is used to determine whether lookup is being done from within a macro argument. We use ASTScopes to find the innermost scope enclosing that source location, and walk up the scope tree to look for either a freestanding macro expansion scope or an attached macro scope. If we find one of those scopes, that source location is contained within a macro argument, and the appropriate flag is added to the name lookup options to exclude macro-generated names from the name lookup results.

### Arbitrary names at global scope

As part of this, I needed to ban freestanding and peer macro expansions at global scope that can generate arbitrary names. Introducing arbitrary names means that any lookup into this scope must expand the macro. This is a problem, because resolving the macro can invoke type checking other declarations. If _anything_ the macro depends on performs unqualified name lookup, e.g. type resolution, we'll get circularity errors. It's better to prevent this by banning these macros at global scope if any of the macro candidates introduce arbitrary names.

Resolves rdar://109214279.

hborla committed:
[ASTScope] Rename and generalize AttachedPropertyWrapperScope to CustomAttributeScope
so that it can represent the scope for any custom attribute and its arguments.

This commit is NFC, but CustomAttributeScope is now applicable to attached macros.

hborla committed:
[ASTScope] Add attribute scopes for nominal type and extension scopes.

hborla committed:
[ASTScope] Add APIs for determining whether a given source location is inside
a macro argument.

hborla committed:
[Macros] Always exclude macro expansions in TypeChecker::lookupMacros.

hborla committed:
[NameLookup] Plumb source location arguments through all name lookup APIs.

This source location will be used to determine whether to add a name lookup
option to exclude macro expansions when the name lookup request is constructed.
Currently, the source location argument is unused.

hborla committed:
[NameLookup] Use source locations to determine whether to exclude macro
expansions.

Evaluator::hasActiveResolveMacroRequest is now unused.

hborla committed:
[PreCheckExpr] Don't perform typo correction inside macro arguments.

hborla committed:
[Macros] Add a test that invokes macro expansion through the conformance
checker while type checking a freestanding macro argument in the same scope.

hborla committed:
[Macros] Ban freestanding and peer macro expansions that can produce arbitrary
names at global scope.

Freestanding and peer macros applied at top-level scope cannot introduce
arbitrary names. Introducing arbitrary names means that any lookup
into this scope must expand the macro. This is a problem, because
resolving the macro can invoke type checking other declarations, e.g.
anything that the macro arguments depend on. If _anything_ the macro
depends on performs name unqualified name lookup, e.g. type resolution,
we'll get circularity errors. It's better to prevent this by banning
these macros at global scope if any of the macro candidates introduce
arbitrary names.

hborla committed:
[ASTScope] Don't consider freestanding macro expansions as being part of a
macro argument.

hborla commented on lib/AST/ASTScopeLookup.cpp:
```diff
@@ -675,3 +675,31 @@ std::pair<CaseStmt *, CaseStmt *> ASTScopeImpl::lookupFallthroughSourceAndDest(
+    // macro expansion buffers for freestanding macros are children of
+    // MacroExpansionDeclScope, and child scopes of freestanding macros
+    // are otherwise inside the macro argument.
+    if (scope->getClassName() == "ASTSourceFileScope")
```
Is there a better way to check the `ASTScope` kind?

DougGregor replied:
No, and it looks like you're not the first to lament that this is the only way to check:

```
  // FIXME: Proper isa/dyn_cast support would be better than a string
  // comparison here.
  while (parent->getClassName() == "GenericParamScope")
    parent = parent->getLookupParent().get();
```

hborla commented:
@swift-ci please smoke test

hborla commented on lib/AST/NameLookupRequests.cpp:
```diff
@@ -545,8 +545,10 @@ static UnqualifiedLookupDescriptor excludeMacrosIfNeeded(
     return descriptor;
 
-  auto &evaluator = descriptor.DC->getASTContext().evaluator;
-  if (!evaluator.hasActiveResolveMacroRequest())
```
@DougGregor is there any reason to keep `hasActiveResolveMacroRequest()` around? It's now unused.

DougGregor replied:
No, let's get rid of it! Your approach is far better, and I don't want us accidentally thinking we should use that operation.

fbartho commented:
Re: Using freestanding & arbitrary peer names now being banned at the global level.

Are you sure that’s what you want to do? I feel like plenty of `@Model` style macros for other engines might need to add arbitrary objects to the global scope.

I’m sure I don’t understand the code at this level though, so please disregard if I don’t make much sense.

hborla commented:
>Are you sure that’s what you want to do? I feel like plenty of @Model style macros for other engines might need to add arbitrary objects to the global scope.

This change does not affect the `@Model` macro, the `@Observable` macro, or other similar macros that add members to types.

hborla committed:
[CodeCompletion] Plumb source locations through code completion name lookup.

hborla committed:
[Parser] Reset ASTScopes before performing the IDEInspection second pass.

IDE inspection can delay parsing of particular declarations, so expanding
ASTScopes during the first pass will miss those declarations. Clear any
expanded scopes to force re-expansion during the second pass.

hborla commented:
@swift-ci please smoke test

DougGregor commented on include/swift/AST/Decl.h:
```diff
@@ -3891,7 +3891,7 @@ class NominalTypeDecl : public GenericTypeDecl, public IterableDeclContext {
   /// set of declarations has not been filtered for visibility, nor have
   /// overridden declarations been removed.
-  TinyPtrVector<ValueDecl *> lookupDirect(DeclName name,
+  TinyPtrVector<ValueDecl *> lookupDirect(DeclName name, SourceLoc loc = SourceLoc(),
```
Should we plan to take away that default argument at some point, once we've managed to update all of the callers?

DougGregor commented on include/swift/AST/NameLookup.h:
```diff
@@ -809,6 +809,10 @@ class ASTScope : public ASTAllocated<ASTScope> {
 
+  /// Returns \c true if the given source location is inside an attached
+  /// or freestanding macro argument.
+  static bool isInMacroArgument(SourceFile *sourceFile, SourceLoc loc);
```
Per my previous comment, since we need some level of name lookup to evaluate this predicate, can we move it alongside the name lookup operations? It makes the layering more clear.

DougGregor commented on lib/Sema/PreCheckExpr.cpp:
```diff
@@ -602,8 +602,14 @@ Expr *TypeChecker::resolveDeclRefExpr(UnresolvedDeclRefExpr *UDRE,
-      TypeChecker::performTypoCorrection(DC, UDRE->getRefKind(), Type(),
-                                         lookupOptions, corrections);
+
+      // FIXME: Don't perform typo correction inside macro arguments, because it
```
Argh, this is a tricky one. Maybe typo correction should avoid ever synthesizing declarations or expanding macros.

DougGregor reviewed (approved):
This looks fantastic. The restriction on arbitrary names at global scope is going to be a bit of a shock at first, but I think we can find suitable alternative approaches. I think I'm convincing myself that it's better overall for readability to limit the introduction of globals.

fbartho commented:
> freestanding and peer macros can still add named, prefixed, suffixed, and overloaded names to the global scope.

Ah! This reassures me/addresses my fear. I thought this was a flat ban on any new names (introduced by macros) at global scope.

Thanks for clarifying!

hborla committed:
[NameLookup] Move macro-related name lookup operations into the namelookup
namespace.

This moves the `isInMacroArgument` predicate and `lookupMacros` into `namelookup`.
ASTScope still encapsulates the scope tree and contains the operation to lookup
the enclosing macro scope, which then invokes a callback to determine whether a
potential macro scope is indeed a macro, because answering this question requires
name lookup.

hborla committed:
[Evaluator] Remove hasActiveResolveMacroRequest().

hborla commented:
@swift-ci please smoke test

hborla commented:
@swift-ci please smoke test

hborla commented:
@swift-ci please smoke test Windows

hborla commented:
@swift-ci please smoke test Windows

WowbaggersLiquidLunch commented:
> As part of this, I needed to ban freestanding and peer macro expansions at global scope that can generate arbitrary names.

If I understand it correctly, does this mean something like [the `#gyb` example](https://github.com/apple/swift-evolution/blob/97197b94b7f64147080f491bb9f81fcd912a1575/proposals/0397-freestanding-declaration-macros.md#template-code-generation) in SE-0397 is no longer possible?

hborla commented:
> If I understand it correctly, does this mean something like [the #gyb example](https://github.com/apple/swift-evolution/blob/97197b94b7f64147080f491bb9f81fcd912a1575/proposals/0397-freestanding-declaration-macros.md#template-code-generation) in [SE-0397](https://apple.github.io/swift-evolution/#?proposal=SE-0397) is no longer possible?

That's correct. However, that example is not producing truly arbitrary names, and it could be expressed if a freestanding macro could use `prefixed`, e.g. `prefixed(Int)`, `prefixed(UInt)`, etc. We're working on a proposal revision that 1) specifies the restriction implemented in this PR, and 2) allows freestanding macros to use `prefixed` and `suffixed` so that we don’t lose all of the expressivity of dynamically generating global names in freestanding declaration macros.

WowbaggersLiquidLunch commented:
> We're working on a proposal revision that 1) specifies the restriction implemented in this PR, and 2) allows freestanding macros to use `prefixed` and `suffixed` so that we don’t lose all of the expressivity of dynamically generating global names in freestanding declaration macros.

Ok this seems reasonable to me. At least, `prefixed` and `suffixed` should be sufficient for my own use case.
